src/snowflake/snowpark/_internal/ast/utils.py

```diff
@@ -1494,3 +1495,28 @@ def base64_lines_to_textproto(base64_str: str) -> str:
 def textproto_to_request(textproto_str: str) -> proto.Request:
     request = Parse(textproto_str, proto.Request())
     return request
+
+
+def clear_line_no_in_ast(ast: Any) -> None:
+    """Clear any 'src' information in the statement body."""
```

Comment:
I am ok with a workaround for fixing this atm to unblock the release, however I actually think we may have a product ask to ensure we've have a proper src position story. Hence, mid-term we need to properly test that correct src positions are extracted. Are you planning on fixing the src code issue in another PR?

Reply:
The root cause is from the inspect library which behaves differently for chained code calls.

For example:
```
42:         df = df.select(
43:             when(col("a") == 1, "one")    ############### py3.10 -> start_line: 43
44:             .otherwise("other_one")
45:             .when(col("a") == 2, "two")
46:             .otherwise("other_two")      ################ py3.10 -> start_line: 46 (rather than 43)
47:             .when(col("a") == 3, "three")
48:             .otherwise("other_three")
49:             .when(col("a") == 4, "four")
50:             .otherwise("other_four")
51:             .when(col("a") == 5, "five")
52:             .otherwise("other_five")
53:         )
```

In python 3.9, inspect frame_info returns line 43 for all of the chained calls (lines 43-52).  Where as for python 3.10, inspect frame_info returns the actual correct line for each chained call (call on line 48 would return line 48.)

As you can see, whether this comes up depends on how the df source code was written.  We could also, for instance, change the test code to avoid hitting this issue.  But ultimately, we care about working with any customer source code, so I'd opt to not change the tests to subtle avoid this in some way.  

It seems we need to change to a different implementation (from inspect, have not looked into alternatives but may be there is something out there) or standardize on python 3.10+ (with updated inspect fixing the above line no bugs.).  Let me know if you have other ideas here.

Reply:
Hm, so to me it sounds like as-if Python 3.10 fixed some wrong behavior or improved the behavior. Considering all subsequent versions should have the Python3.10 behavior, I'd say given Python3.9 will at some point expire (EOL: 1st or 2nd quarter 2025, src: https://devguide.python.org/versions/), let's skip that part of the testing for Python3.9 and test only with 3.10+? Wdyt? This seems like the most pragmatic way to me. I do not really see a good way of fixing up 3.9, and it may not be worth it given 3.9 is already nearing its EOL.

I.e., let's call the clear logic here only for 3.9, update the tests/expectations using 3.10?

Reply:
Updated the expectation test file output based on python 3.11 based on offline conversation.